text/0000-fabric-private-chaincode.md

```diff
@@ -0,0 +1,284 @@
+ 
+- New Runtime Environment: In order to create a TEE-enabled runtime for FPC chaincode, the initial FPC implementation uses a Go wrapper around a C++ shim to enable interaction between the Peer and the Enclave, which didn’t support Go code when the project started. We also use a Peer CLI wrapper to hide Enclave-specific operations. This wrapper approach might not be ideal for compatibility with Fabric build and lifecycle management. As described above, we want to use the new External Launcher feature of Fabric 2.0 to address these issues, and hope to work out the details through the RFC process.
+ 
+- Trusted Ledger Validation: In order to make a trusted “oracle” of information on the current world state, we’ve essentially had to replicate the Peer’s existing validation mechanism inside an Enclave, in C++ - an inefficient duplication of functionality that could lead to version synchronization problems. We hope to explore a path toward a modular validation mechanism that would eliminate this redundancy; possibly enabled by recent work on supporting Go inside the TEE.
+ 
+- Attestation-Based Endorsements: In order to integrate cryptographic Attestation of the TEE with Fabric endorsements, we created custom endorsement and validation plugins and include the enclave signature as part of the response payload. This added payload and the need for every Peer to run a custom validation plugin seem undesirable in the long term. Our current plan is to create a custom Membership Service Provider to replace the original mechanism. The new TEE MSP type would treat TEEs as members of an Org, replacing the Peer’s endorsing signature with a TEE Signature. We believe this approach would be more elegant and flexible, and would lay the ground for easier enablement of new TEE types.
+ 
+- Go Plugins for Trusted Ledger and Validation: In order to build FPC without recompiling the Peer, the current implementation makes use of Go Plugins to enable our Trusted Ledger and Validation functions in the Peer. We understand that Go plugins may not be viable long-term; if they were no longer to be supported, we would have to recompile the Peer.
+ 
+The FPC team welcomes the community’s advice on how each of these touch-points to Fabric should be handled going forward, and hope to solidify our plans for each element through the RFC process.
+ 
+ 
+# Drawbacks
```

Comment:
A few of the following limitations are missing (may be I should call it current limitations).

1. The peer cannot run CouchDB as the stateDB. Even if CouchDB is used, `GetQueryResult()` APIs cannot be used as the value is encrypted and CouchDB is not a database that can operate on encrypted data (unlike cryptDB -- https://css.csail.mit.edu/cryptdb/)
2. `GetStateByRange()` `GetStateByRangeWithPagination()` `GetStateByPartialCompositeKey()` `GetStateByPartialCompositeKeyWithPagination() ` are not supported in the chaincode. 
3. Is the private data supported? -- if not, please state in the limitation.
4. Is the key-based endorsement supported? if not, please state in the limitation.
5. Is it possible for the client to take blocks out of the peer, decrypt rwset, and execute analytical queries?

Reply:
I agree with all these **current** limitations. Let's focus on a robust integration of FPC with a minimum of features but make it end2end secure. From there we can start adding the features listed above.